src/bookshelf/src/bksw_wm_sql_object.erl

```diff
@@ -0,0 +1,315 @@
+        {ok, []} ->
+            io:format("Empty chunk sent Chunk ~p ~p~n", [ChunkId, DbFile]),
+            {[], fun() -> send_streamed_body(Ctx) end};
+        {ok, Data} ->
+            TransferState1 = bksw_sql:update_transfer_state(TransferState0, Data, 1),
+            {Data, fun() -> send_streamed_body(Ctx#context{transfer_state = TransferState1}) end};
+        {error, no_connections} ->
+            %% Sleep a bit then retry.
+            %% not pretty, there should be a nicer way of redoing this. At minimum we probably should push this down
+            %% into bksw_sql or sqerl somehow.
+            timer:sleep(?PGSQL_RETRY_INTERVAL),
+            {[], fun() -> send_streamed_body(Ctx) end};
+        {error, _} = Error ->
+            %% This crashes webmachine; we need a smoother way of handling this.
+            error_logger:error_msg("Error occurred during content download: d~p c~p ~p~n", [DataId, ChunkId, Error]),
+            Error
+    end.
+
+fully_read(#context{} = Ctx, []) ->
+    fully_read(send_streamed_body(Ctx), []);
+fully_read({_, done}, Accum) ->
+    lists:reverse(Accum);
+fully_read({error, Error}, Accum) ->
+    error_logger:error_msg("Error occurred during content download: ~p~n", [Error]),
+    lists:reverse(Accum);
+fully_read({Data, Next}, Accum) ->
+    fully_read(Next(), [Data | Accum]).
+
+%% Creates or replaces a file.
+%% Existing files have their data segment replaced.
+%% New files will have no file_id, and neeed a new file name entry
+finalize_maybe_create_file(_Rq, _Ctx,
+                           #db_file{file_id = undefined,
+                                    bucket_name = BucketName,
+                                    name = Name,
+                                    data_id = DataId}) ->
+    case bksw_sql:create_file_link_data(BucketName, Name, DataId) of
+        {ok, _} ->
+            ok;
+        {error, _Reason} ->
+            %% TODO ERROR HANDLING
+            ok
+    end;
+%% If we have a file_id, we're replacing an old file
+finalize_maybe_create_file(_Rq, _Ctx,
+                           #db_file{file_id = FileId,
+                                    data_id = DataId}) ->
+    case bksw_sql:link_file_data(FileId, DataId) of
+        _ ->
+            ok
+            %% TODO ERROR HANDLING
+    end.
+
+%%
+%% Takes a chunk from webmachine, and streams it out to the db
+%%
+%% A future version of this may want to re-chunk things;
+write_streamed_body({Data, done}, Rq0,
+                    #context{entry_md = #db_file{data_id = DataId} = File0,
+                             transfer_state = #file_transfer_state{next_chunk = ChunkId} = TransferState0} = Ctx0 ) ->
+%%    io:format("Context 2 ~p~n", [Ctx0]),
+
+    #file_transfer_state{next_chunk = ChunkCount} = TransferState1 =
+        case Data of
+            <<"">> ->
+                TransferState0;
+            Data ->
+                %% TODO: This can fail; we need some method for managing failure here.
```

Comment:
This is a real issue, if there's no pooler connection we drop data on the floor; we should at least detect and return an error code so that the user can retry.